pkg/chains/storage/grafeas/grafeas.go

```diff
@@ -0,0 +1,379 @@
+	commonpb "github.com/grafeas/grafeas/proto/v1beta1/common_go_proto"
+	pb "github.com/grafeas/grafeas/proto/v1beta1/grafeas_go_proto"
+	"github.com/pkg/errors"
+	"github.com/tektoncd/chains/pkg/artifacts"
+	"github.com/tektoncd/chains/pkg/chains/formats"
+	"github.com/tektoncd/chains/pkg/config"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	StorageBackendGrafeas = "grafeas"
+	projectNameFormat     = "projects/%s"
+	noteNameFormat        = "projects/%s/notes/%s"
+)
+
+// Backend is a storage backend that stores signed payloads in the storage that
+// is built on the top of grafeas i.e. container analysis.
+type Backend struct {
+	logger *zap.SugaredLogger
+	tr     *v1beta1.TaskRun
+	client pb.GrafeasV1Beta1Client
+	cfg    config.Config
+}
+
+// NewStorageBackend returns a new Grafeas StorageBackend that stores signatures in a Grafeas server
+func NewStorageBackend(ctx context.Context, logger *zap.SugaredLogger, tr *v1beta1.TaskRun, cfg config.Config) (*Backend, error) {
+	// build connection through grpc
+	// implicit uses Application Default Credentials to authenticate.
+	// Requires `gcloud auth application-default login` to work locally
+	creds, err := oauth.NewApplicationDefault(ctx, "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: make grafeas server configurable including checking if hostname is trusted
+	server := "dns:///containeranalysis.googleapis.com"
+
+	conn, err := grpc.Dial(server,
+		grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{})),
+		grpc.WithDefaultCallOptions(grpc.PerRPCCredentials(creds)),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// connection client
+	client := pb.NewGrafeasV1Beta1Client(conn)
+
+	// create backend instance
+	return &Backend{
+		logger: logger,
+		tr:     tr,
+		client: client,
+		cfg:    cfg,
+	}, nil
+}
+
+// StorePayload implements the storage.Backend interface.
+func (b *Backend) StorePayload(ctx context.Context, rawPayload []byte, signature string, opts config.StorageOpts) error {
+	// We only support simplesigning for OCI images, and in-toto for taskrun.
+	if opts.PayloadFormat == formats.PayloadTypeTekton || opts.PayloadFormat == formats.PayloadTypeProvenance {
+		return errors.New("Grafeas storage backend only supports for OCI images and in-toto attestations")
+	}
+
+	// Check if projectID is configured. If not, stop and return an error
+	if b.cfg.Storage.Grafeas.ProjectID == "" {
+		return errors.New("Project ID needs to be configured!")
+	}
+
+	// check if noteID is configured. If not, we give it a name as `tekton-<namespace>`
+	if b.cfg.Storage.Grafeas.NoteID == "" {
+		generatedNoteID := fmt.Sprintf("tekton-%s", b.tr.GetNamespace())
```

Comment:
Hi @priyawadhwa ,
Thanks for proposing this. The current behaviour is that we want to use a single note (either configured by user or this auto-created one) to link to all occurrences created for all taskruns. 
If there are multiple TaskRuns in the same namespace, the note will not be overrode. Instead, it will return AlreadyExistsError when creating an existing note.

Reply:
thanks for explaining!

Reply:
For the integration test, how about we add that in another pr considering the size of the current one?

Reply:
sgtm